bftengine/include/bftengine/CryptoManager.hpp

```diff
@@ -38,28 +38,38 @@ class CryptoManager {
   }
 
  private:
-  CryptoManager(const ReplicaConfig* config, Cryptosystem* cryptoSys) {
+  CryptoManager(const ReplicaConfig* config, Cryptosystem* cryptoSys)
+      : f_{config->fVal}, c_{config->cVal}, numSigners_{config->numReplicas} {
     multiSigCryptoSystem_.reset(cryptoSys);
+    init();
+  }
+
+  void init() {
     thresholdSigner_.reset(multiSigCryptoSystem_->createThresholdSigner());
-    thresholdVerifierForSlowPathCommit_.reset(
-        multiSigCryptoSystem_->createThresholdVerifier(config->fVal * 2 + config->cVal + 1));
-    thresholdVerifierForCommit_.reset(
-        multiSigCryptoSystem_->createThresholdVerifier(config->fVal * 3 + config->cVal + 1));
-    thresholdVerifierForOptimisticCommit_.reset(multiSigCryptoSystem_->createThresholdVerifier(config->numReplicas));
-
-    // TODO [TK] logic for loading keys from reserved pages
-    //    // We want to generate public key for n-out-of-n case
-    //    multiSigCryptoSystem_.reset(new Cryptosystem(MULTISIG_BLS_SCHEME,
-    //                                                 "BN-P254",
-    //                                                 config->numReplicas,
-    //                                                 config->numReplicas));
+    thresholdVerifierForSlowPathCommit_.reset(multiSigCryptoSystem_->createThresholdVerifier(f_ * 2 + c_ + 1));
+    thresholdVerifierForCommit_.reset(multiSigCryptoSystem_->createThresholdVerifier(f_ * 3 + c_ + 1));
+    thresholdVerifierForOptimisticCommit_.reset(multiSigCryptoSystem_->createThresholdVerifier(numSigners_));
+  }
+
+  void updateMultisigKeys(const std::string& secretKey, const std::string& verificationKey) {
+    multiSigCryptoSystem_->updateKeys(secretKey, verificationKey);
+    init();
+  }
+
+  void updateVerificationKeyForSigner(const std::string& verificationKey, const std::uint16_t& signerIndex) {
+    multiSigCryptoSystem_->updateVerificationKey(verificationKey, signerIndex);
+    init();
```

Comment:
Can this method get a boolean `init` , in order to select whether to init on update, since most of the time I'll need to update multiple public keys and call it multiple times.

Reply:
You always need to do init on update as you don't know if there will be another one.

Reply:
I know if multiple public keys need to be updated at once, in that case, it's better to call init once

Reply:
It's a rare event we can not worry about performance here, we can update one by one